ql/src/java/org/apache/hadoop/hive/ql/optimizer/physical/NullScanTaskDispatcher.java

```diff
@@ -93,94 +91,109 @@ private String getAliasForTableScanOperator(MapWork work,
-      aliasesAffected.removeAll(allowed);
-      if (aliasesAffected.isEmpty()) {
-        work.removePathToAlias(path);
-        work.removePathToPartitionInfo(path);
-      }
-    }
+  private void addCandidatePath(Map<Path, Collection<String>> aliasesForPath, Path path, String alias) {
+    Collection<String> aliases = aliasesForPath.computeIfAbsent(path, k -> new ArrayList<>());
+    aliases.add(alias);
   }
 
-  private void processAlias(MapWork work, Set<TableScanOperator> tableScans) {
-    Set<String> aliases = new HashSet<>();
+  private void processTableScans(MapWork work, Set<TableScanOperator> tableScans) {
+    Map<Path, Boolean> managedEmptyPathMap = new HashMap<>();
+    Map<Path, Collection<String>> candidatePathsToAliases = new HashMap<>();
+    Map<String, Boolean> aliasTypeMap = new HashMap<>();
+    Map<String, Map<Path, Partition>> allowedAliasesToPartitions = new HashMap<>();
     for (TableScanOperator tso : tableScans) {
       // use LinkedHashMap<String, Operator<? extends OperatorDesc>>
       // getAliasToWork() should not apply this for non-native table
       if (tso.getConf().getTableMetadata().getStorageHandler() != null) {
         continue;
       }
       String alias = getAliasForTableScanOperator(work, tso);
-      aliases.add(alias);
+      boolean isManagedTable = !MetaStoreUtils.isExternalTable(tso.getConf().getTableMetadata().getTTable());
+      aliasTypeMap.put(alias, isManagedTable);
+      allowedAliasesToPartitions.put(alias, getPathToPartitionMap(alias, tso));
       tso.getConf().setIsMetadataOnly(true);
     }
-    // group path alias according to work
-    Map<Path, List<String>> candidates = new HashMap<>();
+
     for (Path path : work.getPaths()) {
-      List<String> aliasesAffected = work.getPathToAliases().get(path);
-      if (CollectionUtils.isNotEmpty(aliasesAffected)) {
-        candidates.put(path, aliasesAffected);
+      List<String> aliases = work.getPathToAliases().get(path);
+      for (String alias: aliases) {
+        Map<Path, Partition> pathToPartitionMap = allowedAliasesToPartitions.get(alias);
+        if (pathToPartitionMap == null) {
+          continue;
+        }
+        addCandidatePath(candidatePathsToAliases, path, alias);
+        Partition partitionObject = pathToPartitionMap.get(path);
+        Map<String, String> partitionParameters = partitionObject != null ? partitionObject.getParameters() : null;
+        boolean isManagedTable = aliasTypeMap.get(alias);
+        if (isManagedTable && partitionParameters != null && StatsSetupConst.areBasicStatsUptoDate(partitionParameters)) {
+          long rwCount = Long.parseLong(partitionParameters.get(StatsSetupConst.ROW_COUNT));
+          if (rwCount == 0) {
+            managedEmptyPathMap.put(path, true);
+          } else {
+            managedEmptyPathMap.put(path, false);
+          }
+        }
+      }
+    }
+
+    for (Entry<Path, Collection<String>> entry : candidatePathsToAliases.entrySet()) {
```

Comment:
Wouldn't this have the same issue for external tables (e.g text?). For

Reply:
Are you referring to performance issue? Yes this PR does not change anything for external tables. We cannot rely on statistics for external tables. I have another patch ready (need to test first) that deals with external tables. We'll run listStatus calls in parallel for external tables.

Reply:
Thanks for sharing the details.  HIVE-24380 is what you are referring to I believe.

Reply:
Yes